### ISSUE TITLE
Migrate to Kotlin

### ISSUE THREAD

**Issue by gesellix** [docker-client/docker-compose-v3#1]
My philosophy behind the [Docker client](https://github.com/gesellix/docker-client) (and as such this repo, too) is to keep the Docker facing side as flexible as possible. Types might be less convenient, so I have no idea whether Kotlin is counter productive. That's what I want to learn and gain some experience.

My main motivation to introduce Kotlin is the library's client facing side of the libraries. A strongly typed api certainly helps to use a library given a capable IDE. That way, the lack of documentation might be compensated to a certain degree.

Things to keep in mind:

- There's no `static` modifier in Kotlin. `static someMethod(){}` could be converted to `@JvmStatic fun someMethod()`, but it would also be valid to simply remove the `static` modifier and make `someMethod()` an instance method.
- We should focus on the production code first. Test code can also be converted to Kotlin, but with low priority. I'm a fan of the Groovy Spock framework and there's no equivalent in Kotlin, yet (I didn't find any). [Spek](http://spekframework.org/) seems to be the most similar alternative as of today.

See https://github.com/docker-client/docker-filesocket/commit/620cc55d929500da7b72565d5355132c8e0f777b for a first try at converting from Groovy to Kotlin.

**Comment by gesellix** [comment]
Hey @shekhargulati, I have no idea how to start exactly with the migration. If you have any specific idea, just tell me! I would start with the low level classes and work from bottom to up. In this actual library the `de.gesellix.docker.compose.interpolation` should be a good starting point, while `de.gesellix.docker.compose.ComposeFileReader` could be the final step.

The tests are no real unit tests, they are inspired/copied from the original Docker codebase and as such are quite high level. I guess that helps with the migration and ensures we don't completely break everything, so I would convert the tests only after the production code.

**Comment by gesellix** [comment]
Finished with https://github.com/docker-client/docker-compose-v3/commit/d1ba6ee03468f324bd26023056465cc9e08e4459

**Comment by gesellix** [comment]
Follow-up in the docker-client: https://github.com/gesellix/docker-client/issues/44
